File: scripts/commands/commands/universalis.py
from lib.localize import translate_command
from option.ephemeral import ephemeral


async def create_universalis_command():
    command = {
        "name": "universalis",
        "name_localizations": {},
        "description": "Get Final Fantasy XIV market board listings",
        "description_localizations": {},
        "type": 1,
        "skip_localization": True,
        "options": (
            {
                "name": "world",
                "name_localizations": {},
                "description": (
                    "Get Final Fantasy XIV market board listings for a world"
                ),
                "description_localizations": {},
                "options": (
                    {
                        "name": "world",
                        "name_localizations": {},
                        "description": "The world to search in",
                        "description_localizations": {},
                        "required": True,
                        "type": 3,
                    },
                    {
                        "name": "item",
                        "name_localizations": {},
                        "description": "The item to search for",
                        "description_localizations": {},
                        "required": True,
                        "type": 3,
                    },
                    {
                        "name": "high quality",
                        "name_localizations": {},
                        "description": (
                            "Only show high quality market board listings or vice versa"
                        ),
                        "description_localizations": {},
                        "type": 5,
                    },
                    ephemeral,
                ),
                "type": 1,
            },
            {
                "name": "datacenter",
                "name_localizations": {},
                "description": (
                    "Get Final Fantasy XIV market board listings for a datacenter"
                ),
                "description_localizations": {},
                "options": (
                    {
                        "name": "datacenter",
                        "name_localizations": {},
                        "description": "The datacenter to search in",
                        "description_localizations": {},
                        "required": True,
                        "type": 3,
                        "choices": (
                            {"name": "Elemental", "value": "Elemental"},
                            {"name": "Gaia", "value": "Gaia"},
                            {"name": "Mana", "value": "Mana"},
                            {"name": "Aether", "value": "Aether"},
                            {"name": "Primal", "value": "Primal"},
                            {"name": "Chaos", "value": "Chaos"},
                            {"name": "Light", "value": "Light"},
                            {"name": "Crystal", "value": "Crystal"},
                            {"name": "Materia", "value": "Materia"},
                            {"name": "Meteor", "value": "Meteor"},
                            {"name": "Dynamis", "value": "Dynamis"},
                        ),
                    },
                    {
                        "name": "item",
                        "name_localizations": {},
                        "description": "The item to search for",
                        "description_localizations": {},
                        "required": True,
                        "type": 3,
                    },
                    {
                        "name": "high quality",
                        "name_localizations": {},
                        "description": (
                            "Only show high quality market board listings or vice versa"
                        ),
                        "description_localizations": {},
                        "type": 5,
                    },
                    ephemeral,
                ),
                "type": 1,
            },
            {
                "name": "region",
                "name_localizations": {},
                "description": (
                    "Get Final Fantasy XIV market board listings for a region"
                ),
                "description_localizations": {},
                "options": (
                    {
                        "name": "region",
                        "name_localizations": {},
                        "description": "The region to search in",
                        "description_localizations": {},
                        "required": True,
                        "type": 3,
                        "choices": (
                            {
                                "name": "Japan",
                                "value": "Japan",
                                "name_localizations": {},
                            },
                            {
                                "name": "North America",
                                "value": "NorthAmerica",
                                "name_localizations": {},
                            },
                            {
                                "name": "Europe",
                                "value": "Europe",
                                "name_localizations": {},
                            },
                            {
                                "name": "Oceania",
                                "value": "Oceania",
                                "name_localizations": {},
                            },
                        ),
                    },
                    {
                        "name": "item",
                        "name_localizations": {},
                        "description": "The item to search for",
                        "description_localizations": {},
                        "required": True,
                        "type": 3,
                    },
                    {
                        "name": "high quality",
                        "name_localizations": {},
                        "description": (
                            "Only show high quality market board listings or vice versa"
                        ),
                        "description_localizations": {},
                        "type": 5,
                    },
                    ephemeral,
                ),
                "type": 1,
            },
        ),
    }

    return translate_command(command)

File: commands/src/commands/Universalis.kt
package cloud.drakon.dynamisbot.commands

import cloud.drakon.dynamisbot.buildLocalizationMap
import cloud.drakon.dynamisbot.commandName
import cloud.drakon.dynamisbot.ephemeralCommand
import cloud.drakon.dynamisbot.lib.discord.applicationcommand.ApplicationCommand
import cloud.drakon.dynamisbot.lib.discord.applicationcommand.ApplicationCommandOption
import cloud.drakon.dynamisbot.lib.discord.applicationcommand.ApplicationCommandOptionChoiceString

private const val name = "Universalis"
private const val description = "Get Final Fantasy XIV market board listings"

internal suspend fun universalisCommand() = ApplicationCommand(
    type = 1,
    name = name.commandName(),
    nameLocalizations = name.buildLocalizationMap(true),
    description = description,
    descriptionLocalizations = description.buildLocalizationMap(),
    options = buildList {
        arrayOf("datacenter", "region").forEach {
            val name = it.commandName()
            val description = "Get Final Fantasy XIV market board listings for a $it"

            add(
                ApplicationCommandOption(
                    type = 1,
                    name = name,
                    nameLocalizations = it.buildLocalizationMap(true),
                    description = description,
                    descriptionLocalizations = description.buildLocalizationMap(),
                    options = buildList {
                        var description = "The $it to search in"

                        add(
                            ApplicationCommandOption(
                                type = 3,
                                name = name,
                                nameLocalizations = it.buildLocalizationMap(true),
                                description = description,
                                descriptionLocalizations = description.buildLocalizationMap(),
                                choices = when (it) {
                                    "datacenter" -> buildList {
                                        arrayOf(
                                            "Elemental",
                                            "Gaia",
                                            "Mana",
                                            "Meteor",
                                            "Chaos",
                                            "Light",
                                            "Aether",
                                            "Crystal",
                                            "Dynamis",
                                            "Primal",
                                            "Materia",
                                            "NA Cloud DC (Beta)",
                                        ).forEach {
                                            add(ApplicationCommandOptionChoiceString(name = it, value = it))
                                        }
                                    }

                                    "region" -> buildList {
                                        arrayOf("Japan", "Europe", "North America", "Oceania").forEach {
                                            add(
                                                ApplicationCommandOptionChoiceString(
                                                    name = it,
                                                    value = it.replace(" ", "-")
                                                )
                                            )
                                        }
                                    }

                                    else -> null
                                }
                            )
                        )

                        var name = "item"
                        description = "The item to search for"

                        add(
                            ApplicationCommandOption(
                                type = 3,
                                name = name.commandName(),
                                nameLocalizations = name.buildLocalizationMap(true),
                                description = description,
                                descriptionLocalizations = description.buildLocalizationMap(),
                            )
                        )

                        name = "High Quality"
                        description = "Show only High Quality market board listings or vice versa"

                        add(
                            ApplicationCommandOption(
                                type = 5,
                                name = name.commandName(),
                                nameLocalizations = name.buildLocalizationMap(true),
                                description = description,
                                descriptionLocalizations = description.buildLocalizationMap(),
                            )
                        )
                        add(ephemeralCommand())
                    }
                )
            )
        }
    }
)
